x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CheckNotDataStreamWriteIndexStep.java

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.Index;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Some actions cannot be executed on a data stream's write index (eg. `searchable-snapshot`). This step checks if the managed index is
+ * part of a data stream, in which case it will check it's not the write index. If the managed index is the write index of a data stream
+ * this step will wait until that's not the case (ie. rolling over the data stream will create a new index as the data stream's write
+ * index and this step will be able to complete)
+ */
+public class CheckNotDataStreamWriteIndexStep extends ClusterStateWaitStep {
+
+    public static final String NAME = "check-no-write-index";
+
+    private static final Logger logger = LogManager.getLogger(CheckNotDataStreamWriteIndexStep.class);
+
+    CheckNotDataStreamWriteIndexStep(StepKey key, StepKey nextStepKey) {
+        super(key, nextStepKey);
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        Metadata metadata = clusterState.metadata();
+        IndexMetadata indexMetadata = metadata.index(index);
+        String indexName = index.getName();
+
+        if (indexMetadata == null) {
+            String errorMessage = String.format(Locale.ROOT, "[%s] lifecycle action for index [%s] executed but index no longer exists",
+                getKey().getAction(), indexName);
+            // Index must have been since deleted
+            logger.debug(errorMessage);
+            return new Result(false, new Info(errorMessage));
+        }
+
+        String policyName = indexMetadata.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+        IndexAbstraction indexAbstraction = clusterState.metadata().getIndicesLookup().get(indexName);
+        assert indexAbstraction != null : "invalid cluster metadata. index [" + indexName + "] was not found";
+        IndexAbstraction.DataStream dataStream = indexAbstraction.getParentDataStream();
+        if (dataStream != null) {
+            assert dataStream.getWriteIndex() != null : dataStream.getName() + " has no write index";
+            if (dataStream.getWriteIndex().getIndex().getName().equals(indexName)) {
+                String errorMessage = String.format(Locale.ROOT, "index [%s] is the write index for data stream [%s]. " +
+                    "stopping execution of lifecycle [%s]. manually rolling over the index will resume the execution of the policy as " +
+                    "the index will not be the data stream's write index anymore", indexName, dataStream.getName(), policyName);
```

Comment:
I think calling it "stopping" might be a little confusing, I would suggest something like:

"index [.ds-foo-0001] is the write index for data stream [foo], pausing ILM execution until this index is no longer the write index for the data stream via manual or automated rollover"

What do you think?

Reply:
sure, good point